rust/theoros/src/handlers/websocket/subscribe_to_calldata.rs

```diff
@@ -0,0 +1,354 @@
+    },
+    tokio::sync::{broadcast::Receiver, watch},
+};
+
+#[derive(Clone)]
+pub struct DataFeedClientConfig {}
+
+pub struct WsState {
+    pub subscriber_counter: AtomicUsize,
+}
+
+#[allow(clippy::new_without_default)]
+impl WsState {
+    pub fn new() -> Self {
+        Self { subscriber_counter: AtomicUsize::new(0) }
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(tag = "type")]
+enum ClientMessage {
+    #[serde(rename = "subscribe")]
+    Subscribe { ids: Vec<String>, chain_name: EvmChainName },
+    #[serde(rename = "unsubscribe")]
+    Unsubscribe { ids: Vec<String> },
+}
+
+#[derive(Serialize, Debug, Clone)]
+#[serde(tag = "type")]
+enum ServerMessage {
+    #[serde(rename = "response")]
+    Response(ServerResponseMessage),
+    #[serde(rename = "data_feed_update")]
+    DataFeedUpdate { data_feed: RpcDataFeed },
+}
+
+#[derive(Serialize, Debug, Clone)]
+#[serde(tag = "status")]
+enum ServerResponseMessage {
+    #[serde(rename = "success")]
+    Success,
+    #[serde(rename = "error")]
+    Err { error: String },
+}
+
+pub async fn ws_route_handler(
+    ws: WebSocketUpgrade,
+    AxumState(state): AxumState<AppState>,
+    ConnectInfo(_): ConnectInfo<SocketAddr>,
+) -> impl IntoResponse {
+    ws.max_message_size(MAX_CLIENT_MESSAGE_SIZE).on_upgrade(move |socket| websocket_handler(socket, state))
+}
+
+#[tracing::instrument(skip(stream, state))]
+async fn websocket_handler(stream: WebSocket, state: AppState) {
+    let ws_state = state.ws.clone();
+
+    // TODO: add new connection to metrics
+
+    let (sender, receiver) = stream.split();
+
+    let feeds_receiver = state.storage.feeds_channel.subscribe();
+
+    let id = ws_state.subscriber_counter.fetch_add(1, Ordering::SeqCst);
+
+    let mut subscriber = Subscriber::new(id, Arc::new(state.clone()), feeds_receiver, receiver, sender);
+
+    subscriber.run().await;
+}
+
+pub type SubscriberId = usize;
+
+pub struct Subscriber {
+    id: SubscriberId,
+    closed: bool,
+    state: Arc<AppState>,
+    feeds_receiver: Receiver<CheckpointMatchEvent>,
+    receiver: SplitStream<WebSocket>,
+    sender: SplitSink<WebSocket, Message>,
+    data_feeds_with_config: HashMap<String, DataFeedClientConfig>,
+    active_chain: EvmChainName,
+    ping_interval: tokio::time::Interval,
+    exit: watch::Receiver<bool>,
+    responded_to_ping: bool,
+}
+
+impl Subscriber {
+    pub fn new(
+        id: SubscriberId,
+        state: Arc<AppState>,
+        feeds_receiver: Receiver<CheckpointMatchEvent>,
+        receiver: SplitStream<WebSocket>,
+        sender: SplitSink<WebSocket, Message>,
+    ) -> Self {
+        Self {
+            id,
+            closed: false,
+            state,
+            feeds_receiver,
+            receiver,
+            sender,
+            data_feeds_with_config: HashMap::new(),
+            active_chain: DEFAULT_ACTIVE_CHAIN,
+            ping_interval: tokio::time::interval(PING_INTERVAL_DURATION),
+            exit: crate::EXIT.subscribe(),
```

Comment:
Can we store `EXIT` in the state rather than having a static?

Reply:
it is justified as to why it's static in the comment, way easier to use as such